File: solutions/55/main.cpp
#include <iostream>
#include <vector> 
#include <gtest/gtest.h> 

using namespace std; 

class Solution {
public:
    bool canJump(vector<int>& nums) {

        // initialize dp table 
        vector<int> dp(nums.size(), 0);

        for (int i = 1; i < nums.size(); ++i)
        {
            dp[i] = max(dp[i - 1], nums[i - 1]) - 1; 
            if (dp[i] < 0)
                return false; 
        }

        return true; 
    }
};

TEST(Test55, SimpleTest)
{
    vector<int> nums = {2,3,1,1,4}; 
    Solution s; 
    EXPECT_TRUE(s.canJump(nums)) << "Wrong answer"; 
}

int main()
{
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}

File: solutions/55/main.py
from typing import List 

class Solution:
    def canJump(self, nums: List[int]) -> bool:
        maxRange = 0
        i = 0
        while i <= maxRange and i < len(nums):
            maxRange = max(maxRange, nums[i]+i)
            i += 1
        return maxRange >= len(nums)-1

if __name__ == "__main__":
    mySol = Solution()
    nums = [2,3,1,1,4]
    assert mySol.canJump(nums)
